src/main/java/com/slack/kaldb/metadata/zookeeper/CachedMetadataStoreImpl.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * <p>This class also caches nested nodes. However, the root nodes should also have a metadata
+ * object in the data. Further, this cache uses node names as keys, so nested nodes can't have the
+ * same name. This limitation is fine for our current use case. If this becomes a problem switch to
+ * a relative path.
+ *
+ * <p>Currently, the cache is not cleared when a ZK server starts and stops which could be a bug.
+ * But it's fine for now, since we may terminate and restart the process when ZK is unavailable.
+ *
+ * <p>TODO: Cache is refreshed when a ZK server stops/restarts.
+ */
+public class CachedMetadataStoreImpl<T extends KaldbMetadata> implements CachedMetadataStore<T> {
+  private final StandardListenerManager<CachedMetadataStoreListener> listenerContainer =
+      StandardListenerManager.standard();
+  private final AtomicReference<State> state = new AtomicReference<>(State.LATENT);
+  private final CuratorCacheBridge cache;
+
+  private final ConcurrentMap<String, T> instances = Maps.newConcurrentMap();
+
+  private final EnsureContainers ensureContainers;
+  private final CountDownLatch initializedLatch = new CountDownLatch(1);
+  private final MetadataSerializer<T> metadataSerde;
+
+  private enum State {
+    LATENT,
+    STARTED,
+    STOPPED
+  }
+
+  private static ExecutorService convertThreadFactory(ThreadFactory threadFactory) {
+    Preconditions.checkNotNull(threadFactory, "threadFactory cannot be null");
+    return Executors.newSingleThreadExecutor(threadFactory);
+  }
+
+  CachedMetadataStoreImpl(
+      String path,
+      MetadataSerializer<T> metadataSerde,
+      CuratorFramework curator,
+      ThreadFactory threadFactory) {
+    this(path, metadataSerde, curator, convertThreadFactory(threadFactory));
+  }
+
+  CachedMetadataStoreImpl(
+      String path,
+      MetadataSerializer<T> metadataSerde,
+      CuratorFramework curator,
+      ExecutorService executorService) {
+    Preconditions.checkNotNull(path, "name cannot be null");
+    Preconditions.checkNotNull(metadataSerde, "metadata serializer cannot be null");
+    Preconditions.checkNotNull(curator, "curator framework cannot be null");
+    this.metadataSerde = metadataSerde;
+
+    // Create a curator cache but don't store any data in it since CacheStorage only allows
+    // storing data as a byte array. Instead use the curator cache implementation for
+    // managing persistent watchers and other admin tasks. Instead add a listener which would
+    // cache the data locally as a POJO using a serializer. In future, this also allows us to store
+    // the data in a custom data structure other than a hash table. Currently, if we lose a ZK
+    // connection the cache will grow stale but this class is oblivious of it.
+    // TODO: Add a mechanism to detect a stale cache indicate that a cache is stale.
+    cache =
+        CuratorCache.bridgeBuilder(curator, path)
+            .withExecutorService(executorService)
```

Comment:
```
        if ( !forceTreeCache && Compatibility.hasPersistentWatchers() )
        {
            if ( executorService != null )
            {
                LoggerFactory.getLogger(getClass()).warn("CuratorCache does not support custom ExecutorService");
            }
            CuratorCacheStorage storage = cacheData ? CuratorCacheStorage.standard() : CuratorCacheStorage.dataNotCached();
            return new CuratorCacheImpl(client, storage, path, options, null);
        }
```

If you see the build code we don't need an executor service

Reply:
Why not instantiate CuratorCacheImpl directly since we know we will be using ZK 3.6+ ?

Reply:
I preferred the bridge builder since the it is also backwards compatible with older versions of ZK also. Several recipes also use this pattern so it is widely supported and safe to do so.